### ISSUE TITLE
CreateCustomTokenAsync throws Exception: "Method Not Found System.String.TrimEnd"

### ISSUE THREAD

**Issue by mikeperrett** [firebase/firebase-admin-dotnet#65]
@hiranya911 

The unit tests succeed when calling JwtUtils::UrlSafeBase64Encode around line 89 with this code:
return base64Value.TrimEnd('=').Replace('+', '-').Replace('/', '_');

I don't understand how that is so, since System.String.TrimEnd takes as a parameter a char array and you are passing a single char: '='

After recompiling the project with this change to "TrimEnd", it now works as expected in my project:
return base64Value.**TrimEnd(new[] { '=' })**.Replace('+', '-').Replace('/', '_');

Without my fix, here is the stack trace I get when calling CreateCustomTokenAsync:

`Method not found: 'System.String System.String.TrimEnd(Char)'.
at FirebaseAdmin.Auth.JwtUtils.UrlSafeBase64Encode(Byte[] bytes)   at FirebaseAdmin.Auth.JwtUtils.Encode(Object obj)   at FirebaseAdmin.Auth.JwtUtils.<CreateSignedJwtAsync>d__3.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at FirebaseAdmin.Auth.FirebaseTokenFactory.<CreateCustomTokenAsync>d__8.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at FirebaseAdmin.Auth.FirebaseAuth.<CreateCustomTokenAsync>d__13.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at FirebaseAdmin.Auth.FirebaseAuth.<CreateCustomTokenAsync>d__11.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at FirebaseAdmin.Auth.FirebaseAuth.<CreateCustomTokenAsync>d__10.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()   at UserQuery.<GetUserTokenAsync>d__11.MoveNext() in C:\Users\mikep\AppData\Local\Temp\LINQPad5\_jkekndcd\query_ghgvxj.cs:line 151
--`

**Comment by mikeperrett** [comment]
@hiranya911 thank you! Indeed the code is correctly written. When I compile firebase-admin-dotnet locally and reference the locally compiled dll, everything works as expected. So, at the very least, I can assume that there is something outside of the code itself, as you've pointed out above.

My environment is Windows 10, Visual Studio 2017 Pro, I tried changing my project to different versions of .net at 4.5 and greater all with the same exception. 
Although I have not yet got to the root of the problem in my environment, I closed the PR.

**Comment by hiranya911** [comment]
I'm going to close this since I don't see there's a bug to be fixed. Feel free to reopen with more details if something needs to be done.

**Comment by ruaridhnewman** [comment]
Hi @hiranya911 , I found a similar issue when using `VerifyIdTokenAsync`, which throws an error:

 `Method not found: 'System.String[] System.String.Split(Char, System.StringSplitOptions)'.`

when using .NET Framework 4.7. It is broken for releases 1.5.0 and 1.6.0, but works for 1.3.0 and 1.4.0 so something that was released in 1.5.0 seems to have broken it. It works as expected for a .NET Core project, so just seems to be something to do with how .NET Framework 4.7 is compiling the dll's or something.

Other methods such as `GetUserByEmailAsync` work fine, so it looks like it is just something to do with that method.

If you need any other info let me know, but its easily replicated by creating a simple .NET Framework 4.7 Web API project, installing FirebaseAdmin 1.6.0 through nuget, and using the `VerifyIdTokenAsync` method.

This is a critical method for us, so it would be great if it could be looked into quickly! Thanks!

**Comment by hiranya911** [comment]
Nothing has changed in the token verification code since March: https://github.com/firebase/firebase-admin-dotnet/commits/master/FirebaseAdmin/FirebaseAdmin/Auth/FirebaseTokenVerifier.cs

So I assume this is some issue with how the releases are bundled. Does anybody know what's going on here? I see similar issues reported in other projects, but don't see specific resolution: https://answers.unity.com/questions/1478418/custom-dll-missingmethodexception-string-stringspl.html

**Comment by hiranya911** [comment]
I only have access to a Linux/Mono environment at the moment. I tried reproducing the issue there, but that didn't work. Here's what I tried:

ExampleApp.csproj
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net47</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="FirebaseAdmin" Version="1.6.0" />
  </ItemGroup>
</Project>
```

HelloWorld.cs
```
using System;
using System.Threading.Tasks;
using FirebaseAdmin;
using FirebaseAdmin.Auth;
using Google.Apis.Auth.OAuth2;

public class HelloWorld
{

    public static async Task Main(string[] args)
    {
	    FirebaseApp.Create(new AppOptions()
        {
            Credential = GoogleCredential.GetApplicationDefault(),
            ProjectId = "test-project",
        });
        var fakeToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0Ijox"
            + "NTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5";
        try
        {
            var result = await FirebaseAuth.DefaultInstance.VerifyIdTokenAsync(fakeToken);
            Console.WriteLine(result);
        }
        catch (FirebaseException ex)
        {
            Console.WriteLine(ex);
        }
    }
}
```

Compile and run:
```
$ dotnet build /p:OS="Windows_NT" /p:FrameworkPathOverride="/usr/lib/mono/4.7-api/"
$ mono bin/Debug/net47/ExampleApp.exe
```

This fails with the expected exception:
```
FirebaseAdmin.FirebaseException: VerifyIdTokenAsync() expects an ID token, but was given a legacy custom token.
  at FirebaseAdmin.Auth.FirebaseTokenVerifier.VerifyTokenAsync (System.String token, System.Threading.CancellationToken cancellationToken)
```

Can you provide a similar example, that I can try on my machine?

**Comment by jskeet** [comment]
I've reproduced the issue on Windows with @hiranya911's exact project/program file (with an additional LangVersion element for async main).

Now looking into what's going on here.

**Comment by jskeet** [comment]
Okay, so this is the line of code that's broken, in FirebaseTokenVerifier:

```csharp
string[] segments = token.Split('.');
```

For .NET 4.7, that *should* be generating IL like this:

```il
callvirt   instance string[] [mscorlib]System.String::Split(char[])
```
(That's what the same code calls if you put it in the sample app.)

However, the IL in FirebaseAdmin.dll does indeed look like it's trying to call a different method:

```il
callvirt   instance string[] [mscorlib]System.String::Split(char, valuetype [mscorlib]System.StringSplitOptions)
```

It looks like that overload *does* exist in .NET Core 2.0+, but not in .NET Standard 2.0 or in desktop .NET.

So basically, I believe this is an issue with how the package has been compiled. @hiranya911 Do you know about the environment on which this is being built? The target frameworks look okay to me, so if this is built by the regular .NET Core SDK, I'd expect it to be fine.

**Comment by ruaridhnewman** [comment]
Thanks for looking into this so quickly! When I copy across the dlls from a .net core compiled version it works fine, so I agree, looks like something to do with how its compiled.

**Comment by hiranya911** [comment]
Thanks @jskeet  for looking into this. We do build our releases using .NET core. Here's `dotnet --info` output from the build environment:

```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.203
 Commit:    e5bab63eca

Runtime Environment:
 OS Name:     debian
 OS Version:  
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   ~/dotnetsdk-2.2.203/sdk/2.2.203/

Host (useful for support):
  Version: 2.2.4
  Commit:  f95848e524

.NET Core SDKs installed:
  2.2.203 [~/dotnetsdk-2.2.203/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.4 [~/dotnetsdk-2.2.203/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.4 [~/dotnetsdk-2.2.203/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.4 [/~/dotnetsdk-2.2.203/shared/Microsoft.NETCore.App]
```

Now if I remember correct, this machine used to have .NET Core 2.1 until recently. May be upgrading to 2.2 was what caused the issue. I can try compiling some DLLs on this machine and see if that's the issue. 

@jskeet any tooling suggestions for inspecting the IL on Linux environments?

**Comment by jskeet** [comment]
Updating to the 2.2 SDK *shouldn't* have caused this. But given that it targets netstandard1.5 and net45, presumably you're building on a Windows box somewhere? (We can continue this part internally, of course.)

In terms of an equivalent to ildasm but on Linux - I'm afraid I don't know of one that's widely adopted. https://www.nuget.org/packages/dotnet-ildasm/ looks encouraging - I haven't tried it yet, but will do so when I get a chance.

**Comment by jskeet** [comment]
I can confirm that that tool works really simply:

```sh
$ dotnet tool install --global dotnet-ildasm --version 0.9.4
$ dotnet ildasm path/to/dll
```

**Comment by hiranya911** [comment]
I can repro the difference in IL by compiling against different .NET framework versions:

**/usr/lib/mono/4.5/ (default):** IL_000b: callvirt instance string[] string::Split(char, [mscorlib]System.StringSplitOptions)
**/usr/lib/mono/4.5-api/:** IL_001c: callvirt instance string[] string::Split(char[])
**/usr/lib/mono/4.7-api/:** IL_001c: callvirt instance string[] string::Split(char[])

I wonder if something has changed in the mono installation, which may have resulted in this issue. Perhaps we can start building against 4.5-api?

**Comment by hiranya911** [comment]
This should be fixed in the 1.7.0 release. We have moved our release pipelines to Windows with .NET Framework.

**Comment by ruaridhnewman** [comment]
Great, thank you!

**Comment by mikeperrett** [comment]
Awesome!, thanks so much

**Comment by chientranse** [comment]
@hiranya911 Thank you so much!
